Link error for main on Linux x86

this is likely the same bug as #36 but lets track it separately for now.

```
started lowering AST
finished lowering AST
started type resolution
finished type resolution
COLLECT_GCC_OPTIONS='-fsplit-stack' '-v' '-g' '-O0' '-o' 'test' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 as -v --64 -o /tmp/cchYC833.o /tmp/ccPPyd64.s
GNU assembler version 2.34 (x86_64-linux-gnu) using BFD version (GNU Binutils for Ubuntu) 2.34
COMPILER_PATH=/opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/11.0.0/:/opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/11.0.0/:/opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/:/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/:/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/
LIBRARY_PATH=/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/:/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/../../../../lib64/:/lib/x86_64-linux-gnu/:/lib/../lib64/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib64/:/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-fsplit-stack' '-v' '-g' '-O0' '-o' 'test' '-shared-libgcc' '-mtune=generic' '-march=x86-64' '-dumpdir' 'test.'
 /opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/11.0.0/collect2 -plugin /opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/11.0.0/liblto_plugin.so -plugin-opt=/opt/gccrs-dev/libexec/gcc/x86_64-pc-linux-gnu/11.0.0/lto-wrapper -plugin-opt=-fresolution=/tmp/ccBljRM2.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o test /lib/x86_64-linux-gnu/crt1.o /lib/x86_64-linux-gnu/crti.o /opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/crtbegin.o -L/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0 -L/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/../../.. /tmp/cchYC833.o -lm --wrap=pthread_create -lgcc_s -lgcc -lc -lgcc_s -lgcc /opt/gccrs-dev/lib/gcc/x86_64-pc-linux-gnu/11.0.0/crtend.o /lib/x86_64-linux-gnu/crtn.o
/usr/bin/ld: /lib/x86_64-linux-gnu/crt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status```

--- a/gcc/rust/backend/rust-compile-item.h
+++ b/gcc/rust/backend/rust-compile-item.h
@@ -138,10 +138,24 @@ public:
     // convert to the actual function type
     auto compiled_fn_type = TyTyCompile::compile (ctx->get_backend (), fnType);
 
+    unsigned int flags = 0;
+    bool is_main_fn = function.function_name.compare ("main") == 0;
+
+    // if its the main fn or pub visibility mark its as DECL_PUBLIC
+    // please see https://github.com/Rust-GCC/gccrs/pull/137
+    if (is_main_fn || function.has_visibility ())
+      flags |= Backend::function_is_visible;
+
+    std::string asm_name = function.function_name;
+    if (!is_main_fn)
+      {
+	// FIXME need name mangling
+	asm_name = "__" + function.function_name;
+      }
+
     Bfunction *fndecl
       = ctx->get_backend ()->function (compiled_fn_type, function.function_name,
-				       "" /* asm_name */, 0 /* flags */,
-				       function.get_locus ());
+				       asm_name, flags, function.get_locus ());
     ctx->insert_function_decl (function.get_mappings ().get_hirid (), fndecl);
 
     // setup the params

